Assets/Scripts/AttackLogic.cs

```diff
@@ -0,0 +1,175 @@
+    [SerializeField]
+    private double attackTime;
+
+    [Tooltip("The maximum difference in degrees for the attacker between look direction and target direction in order to perform a successful hit")]
+    [SerializeField]
+    private double hitRotationTolerance;
+
+    [Tooltip("Stops attacking the target after a (un-)successful hit")]
+    [SerializeField]
+    private bool resetAfterHit;
+
+
+    public bool PerformingHit { get; private set; }
+
+    // component references
+    private IMovable _movable;
+    
+    private float _timer;
+    // can only hold IDamageables
+    private GameObject _target;
+
+    // ----temporary as animation replacement------
+        private MeshRenderer _gameObjectRenderer;
+        private Material _prevMat;
+        private Material _attackAnimationMaterial;
+    // ---------
+
+    private void Awake()
+    {
+        // init components
+        _movable = GetComponent<IMovable>();
+
+        // -----temp replacement for animation-----
+            _gameObjectRenderer = gameObject.GetComponentInChildren<MeshRenderer>();
+            _prevMat = _gameObjectRenderer.material;
+            _attackAnimationMaterial = new Material(Shader.Find("Standard"));
+            _attackAnimationMaterial.color = Color.yellow;
+        // ----------
+    }
+
+    /// <summary>
+    /// If it has a target, attacks it.
+    /// </summary>
+    private void Update()
+    {
+        if (_target != null)
+        {
+            Attack();
+        }
+    }
+
+    /// <summary>
+    /// Attacks the target. If the target is not in attacking range, chases it. If the target is near
+    /// performs a hit on it. If currently performing a hit, waits for finishing the hit and deals 
+    /// damage if hit was successful. Either resets afterwards or continues chasing / attacking target.
+    /// </summary>
+    private void Attack()
+    {
+        
+
+        float distanceToTarget = Vector3.Distance(_target.transform.position, transform.position);
+        if (distanceToTarget < attackRange && !PerformingHit)
+        {
+            // is in range
+            _movable.StopMoving();
+            // face target
+            if (_movable.FaceTarget(_target, true))
+            {
+                // faces target
+                // perform hit
+                PerformingHit = true;
+
+                // -----temp-----
+                    _prevMat = _gameObjectRenderer.material;
+                
+                    _gameObjectRenderer.material = _attackAnimationMaterial;
+                // ----------
+
+            }
+        }
+        else if (PerformingHit)
+        {
+            bool? success = PerformHit();
+            if (success != null)
+            {
+                // hit animation performed
+                if (success == true)
+                {
+                    // deal damage
+                    IDamageable damageable = (IDamageable)_target.GetComponent<IDamageable>();
```

Comment:
_target is of type GameObject and will be set at some point. Is there a reason why you can't define it as a IDamageable from the beginning? Attacking will happen more often than setting the _target, right?

Reply:
I don't know if setting the target (clicking on the enemy) will happen less often than a successful hit... especially because I know people that rape the mouse button on attack :D

(I need the gameobject so that I have the position of the enemy and that I can calculate the distance.)